File: bindings/java/examples/Streaming.java
import io.vietasr.Pipeline;
import io.vietasr.Session;

import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.IOException;

public final class Streaming {
    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            System.err.println("usage: Streaming <wav-file>");
            System.exit(1);
        }

        WavData wav = readWav(args[0]);
        double durationS = (double) wav.pcm.length / wav.sampleRate;
        System.out.printf("audio: %.2fs @ %d Hz%n", durationS, wav.sampleRate);

        try (Pipeline pipe = Pipeline.preset("transcribe");
             Session session = pipe.stream(wav.sampleRate)) {

            int chunk = wav.sampleRate / 1000 * 320;
            long started = System.nanoTime();
            String lastPartial = "";

            for (int offset = 0; offset < wav.pcm.length; offset += chunk) {
                int end = Math.min(offset + chunk, wav.pcm.length);
                short[] slice = new short[end - offset];
                System.arraycopy(wav.pcm, offset, slice, 0, slice.length);
                session.accept(slice);
                String partial = session.partial().text();
                if (!partial.isEmpty() && !partial.equals(lastPartial)) {
                    System.out.printf("  [%5.2fs] %s%n",
                            (double) end / wav.sampleRate, partial);
                    lastPartial = partial;
                }
            }

            String finalText = session.finalResult().text();
            double elapsed = (System.nanoTime() - started) / 1e9;
            System.out.printf("%nFINAL (%.2fs wall, RTF %.2f):%n%s%n",
                    elapsed, elapsed / durationS, finalText);
        }
    }

    private static final class WavData {
        final short[] pcm;
        final int sampleRate;

        WavData(short[] pcm, int sampleRate) {
            this.pcm = pcm;
            this.sampleRate = sampleRate;
        }
    }

    private static WavData readWav(String path) throws IOException {
        try (DataInputStream in = new DataInputStream(new FileInputStream(path))) {
            byte[] header = new byte[12];
            in.readFully(header);
            if (!"RIFF".equals(new String(header, 0, 4))
                    || !"WAVE".equals(new String(header, 8, 4))) {
                throw new IOException("not a RIFF/WAVE file");
            }
            int sampleRate = 16000;
            int channels = 1;
            int bits = 16;
            while (true) {
                byte[] chunkHeader = new byte[8];
                in.readFully(chunkHeader);
                String id = new String(chunkHeader, 0, 4);
                int size = le32(chunkHeader, 4);
                byte[] body = new byte[size];
                in.readFully(body);
                if ("fmt ".equals(id)) {
                    channels = le16(body, 2);
                    sampleRate = le32(body, 4);
                    bits = le16(body, 14);
                } else if ("data".equals(id)) {
                    if (bits != 16) {
                        throw new IOException("only 16-bit PCM supported");
                    }
                    int total = size / 2;
                    short[] all = new short[total];
                    for (int i = 0; i < total; ++i) {
                        all[i] = (short) le16(body, i * 2);
                    }
                    if (channels == 1) {
                        return new WavData(all, sampleRate);
                    }
                    short[] mono = new short[total / channels];
                    for (int i = 0; i < mono.length; ++i) {
                        int mixed = 0;
                        for (int c = 0; c < channels; ++c) {
                            mixed += all[i * channels + c];
                        }
                        mono[i] = (short) (mixed / channels);
                    }
                    return new WavData(mono, sampleRate);
                }
            }
        }
    }

    private static int le16(byte[] b, int off) {
        return (b[off] & 0xFF) | ((b[off + 1] & 0xFF) << 8);
    }

    private static int le32(byte[] b, int off) {
        return (b[off] & 0xFF) | ((b[off + 1] & 0xFF) << 8)
                | ((b[off + 2] & 0xFF) << 16) | ((b[off + 3] & 0xFF) << 24);
    }
}

File: bindings/rust/examples/streaming.rs
use std::fs::File;
use std::io::Read;
use std::process::exit;
use std::time::Instant;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("usage: streaming <wav-file>");
        exit(1);
    }

    let (pcm, sample_rate) = match read_wav_mono_i16(&args[1]) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("read wav: {e}");
            exit(2);
        }
    };
    let duration_s = pcm.len() as f64 / sample_rate as f64;
    println!("audio: {duration_s:.2}s @ {sample_rate} Hz");

    let pipeline = vietasr::Pipeline::preset("transcribe").expect("preset");
    let mut session = pipeline.stream(sample_rate as f32).expect("stream");

    let chunk = (sample_rate / 1000 * 320) as usize;
    let started = Instant::now();
    let mut last_partial = String::new();

    let mut offset = 0;
    while offset < pcm.len() {
        let end = (offset + chunk).min(pcm.len());
        session.accept_i16(&pcm[offset..end]);
        let partial = session.partial().text();
        if !partial.is_empty() && partial != last_partial {
            let t = end as f64 / sample_rate as f64;
            println!("  [{t:6.2}s] {partial}");
            last_partial = partial;
        }
        offset = end;
    }

    let final_text = session.finalize().text();
    let elapsed = started.elapsed().as_secs_f64();
    println!();
    println!(
        "FINAL ({elapsed:.2}s wall, RTF {:.2}):\n{final_text}",
        elapsed / duration_s
    );
}

fn read_wav_mono_i16(path: &str) -> Result<(Vec<i16>, u32), String> {
    let mut bytes = Vec::new();
    File::open(path)
        .map_err(|e| e.to_string())?
        .read_to_end(&mut bytes)
        .map_err(|e| e.to_string())?;

    if bytes.len() < 12 || &bytes[0..4] != b"RIFF" || &bytes[8..12] != b"WAVE" {
        return Err("not a RIFF/WAVE file".into());
    }

    let mut sample_rate: u32 = 16000;
    let mut channels: u16 = 1;
    let mut bits: u16 = 16;
    let mut offset = 12usize;

    while offset + 8 <= bytes.len() {
        let id = &bytes[offset..offset + 4];
        let size = u32::from_le_bytes(bytes[offset + 4..offset + 8].try_into().unwrap()) as usize;
        let body = offset + 8;
        if id == b"fmt " {
            channels = u16::from_le_bytes(bytes[body + 2..body + 4].try_into().unwrap());
            sample_rate = u32::from_le_bytes(bytes[body + 4..body + 8].try_into().unwrap());
            bits = u16::from_le_bytes(bytes[body + 14..body + 16].try_into().unwrap());
        } else if id == b"data" {
            if bits != 16 {
                return Err("only 16-bit PCM supported".into());
            }
            let total = size / 2;
            let mut all = Vec::with_capacity(total);
            for i in 0..total {
                all.push(i16::from_le_bytes(
                    bytes[body + i * 2..body + i * 2 + 2].try_into().unwrap(),
                ));
            }
            if channels == 1 {
                return Ok((all, sample_rate));
            }
            let ch = channels as usize;
            let mono: Vec<i16> = all
                .chunks(ch)
                .map(|frame| {
                    let sum: i32 = frame.iter().map(|&s| s as i32).sum();
                    (sum / ch as i32) as i16
                })
                .collect();
            return Ok((mono, sample_rate));
        }
        offset = body + size + (size & 1);
    }
    Err("no data chunk".into())
}
